### PR TITLE
compiler/next: Get all but a few primers to compile with new parser

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This is for `compiler/next`.

Get all but a few primers (`chpldoc`, `FFTWlib`, `LAPACKlib`) to
compile with the new frontend via the `--compiler-library-parser`
flag.

Add a few new UAST nodes: `Scan`, `TypeQuery`, and `VarArgFormal`.

Adjust the `TupleDecl` node so that it can also contain `Formal`.

Adjust the `Function::formals()` iterator and accessor to yield
`Decl` instead of `Formal`.

Adjust the conversion routines in `compiler/passes/convert-uast.cpp`
as needed to translate new UAST nodes as well as new combinations
of existing nodes (such as edge cases for array types).

Wire up a bunch of productions in the new parser grammar such as
for scans, varargs, loop expressions with filters, and more.

Remove redundant logic for setting config vars in the production
compiler. Reuse this function in the conversion pass for now.

Add a bunch of TODOs for functionality in the conversion pass
powered by the production compiler that should be rewritten
in the new compiler when possible.

Reviewed by @mppf, @arezaii. Thanks!

TESTING:

- [x] All tests with `linux64`, `COMM=none` to verify config change is OK
- [x] All primers except `FFTWlib`, `LAPACKlib`, `chpldoc` with
        `--compiler-library-parser` flag

FUTURE WORK:

- Handle pragmas and conversion of pragmas (ideally using an approach
  like that in #18619)
- Tackle all remaining TODOEXPR in the the new grammar file
- Pick another larger group of tests to run next
- Add verbosity levels to errors (e.g. `noteWarning`)
- Begin iterating through all production compiler `build` routines
  (used by the old parser) and ensuring that their warnings and
  errors also occur in the new parser
- Consider traversing the tree after parsing (say while assigning
  postorder IDs) and emitting some parse-time errors in order to
  decouple these errors from the parser proper
- Begin to selectively parse the internal modules with the new parser
- Add machinery to parse and look up config var values

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

### CODE REVIEW

File: compiler/next/include/chpl/uast/Scan.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+#include "chpl/queries/Location.h"
+#include "chpl/queries/UniqueString.h"
+#include "chpl/uast/Call.h"
+
+namespace chpl {
+namespace uast {
+
+
+/**
+  This class represents a scan.
```
Comment (arezaii) [inline compiler/next/include/chpl/uast/Scan.h:32]: a little more text would be helpful in this comment

File: compiler/next/include/chpl/uast/VarArgFormal.h [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+ private:
+  int countChildNum_;
+
+  VarArgFormal(ASTList children, UniqueString name, Formal::Intent intent,
+               int8_t typeExpressionChildNum,
+               int8_t countChildNum)
+    : VarLikeDecl(asttags::VarArgFormal, std::move(children),
+                  Decl::DEFAULT_VISIBILITY,
+                  Decl::DEFAULT_LINKAGE,
+                  /*linkageNameChildNum*/ -1,
```
Comment (arezaii) [inline compiler/next/include/chpl/uast/VarArgFormal.h:53]: Could you comment on what the -1 means in this context and the one below?

**Review by arezaii — APPROVED** [review]
Great job getting all the primers working!
I just had a couple of very minor comment requests.

File: compiler/next/include/chpl/uast/VarArgFormal.h (hunk 2 above)
Comment (mppf) [reply compiler/next/include/chpl/uast/VarArgFormal.h:53]: When talking about child numbers (aka index into a vector of children in the uAST tree) we have been using -1 to indicate the absence of that child. (So here specifically it is saying that there is no linkage name child uAST node).

**Pull request merged**